### DISCUSSION TITLE
Aggregate function to get the first array that is not empty

### DISCUSSION THREAD

**Discussion by samuelvl** [ClickHouse/ClickHouse#71487] [Q&A] [answered]
I have a very simple use case, I have a table with an ID and a set of elements stored in an array:

```sql
CREATE TABLE array_test (id Int64, val Array(String)) ENGINE = MergeTree ORDER BY id;
INSERT INTO array_test (id, val) Values (1, null), (1, ['foo']), (1, ['bar']);
INSERT INTO array_test (id, val) Values (2, null), (2, ['bar']), (2, ['foo']);
```

I want to group by ID and get the first array that is not empty for every ID. For example, in the previous example I would like to get `(1, ['foo']), (2, ['bar'])`.

For other simple types I can use the [any aggregate function](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/any) but it doesn't seem compatible with arrays, when I do:

```sql
SELECT id, any(val) FROM array_test GROUP BY id;
```

I get the following result:

```
   ┌─id─┬─any(val)─┐
1. │  2 │ []       │
2. │  1 │ []       │
   └────┴──────────┘
```

Is there any alternative to any for empty arrays?

**Comment by samuelvl** [comment]
I managed to do it with the following query:

```sql
SELECT
    id,
    arrayFirst(x -> x IS NOT NULL AND length(x) > 0, groupArray(val)) as val
FROM array_test
GROUP BY id;
```

**Comment by Algunenano** [comment, marked as answer]
You can use the -If [combinator](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/combinators):

```
SELECT
    id,
    anyIf(val, NOT empty(val))
FROM array_test
GROUP BY id

Query id: f0ab7f51-2ff7-4bb9-9262-4e4e057f32b1

   ┌─id─┬─anyIf(val, not(empty(val)))─┐
1. │  2 │ ['bar']                     │
2. │  1 │ ['foo']                     │
   └────┴─────────────────────────────┘
```

**Reply by samuelvl** [reply]
Awesome, much more simple!
